### PR TITLE
(test) Add scaffolding for better testing

### PR DESCRIPTION
**Pull request by ibacher**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/en-US/docs/frontend-modules/contributing/#contributing-guidelines) label. See existing PR titles for inspiration.
- [ ] My work is based on designs, which are linked or shown either in the Jira ticket or the description below. (See also: [Styleguide](http://om.rs/o3ui))
- [x] My work includes tests or is validated by existing tests.

## Summary

Medium-term plan is to do some refactoring to clean-up the form-engine library a bit and port it over to use O3Forms. In order to do so, first, though, I want to make sure to have more solid test coverage of various edge-cases. This is the first commit in a series as I'm trying to keep them somewhat reviewable (I realize 1300 lines is pushing things).

This PR adds:

* Scaffolding for future tests to build on, namely a few new testing utilities in `/test-support/`, which are excluded from the production build
* Tests for:
  * The `formStateReducer` (the core state tracking across forms)
  * The newly added test utilities
  * Various other utilities that lacked tests 
  * `evaluatePostSubmissionExpression`

It also moves the registry-cache and ast-cache into their own modules, mostly so these can be reset in the test suite without needing to import, e.g., the whole registry or the whole expression evaluation machinery.

### CODE REVIEW

File: src/utils/ast-cache.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import { type compile } from '@openmrs/esm-framework';
```
[the remark this answers is not in the document]
Comment (ibacher) [reply src/utils/ast-cache.ts:1]: This answer is incorrect, at least for recent versions of TS.

File: src/test-support/openmrs-fetch-router.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+export function mockOpenmrsFetchRoutes(routes: FetchRoute[]): Mock {
+  const fetchMock = openmrsFetch as unknown as Mock;
+  routerInstalled = true;
+  fetchMock.mockImplementation(async (url: string, init?: RequestInit) => {
+    const method = (init?.method ?? 'GET').toUpperCase();
+    const route = routes.find((candidate) => {
+      if (candidate.method && candidate.method.toUpperCase() !== method) {
+        return false;
+      }
+      return candidate.match instanceof RegExp ? candidate.match.test(url) : candidate.match(url);
```
[the remark this answers is not in the document]
Comment (ibacher) [reply src/test-support/openmrs-fetch-router.ts]: Done

File: src/test-support/reset.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    // lifecycle teardown only reaches adapters a mounted form registered.
+    ['ObsAdapter.tearDown', () => ObsAdapter.tearDown()],
+    ['OrdersAdapter.tearDown', () => OrdersAdapter.tearDown()],
+    ['EncounterDiagnosisAdapter.tearDown', () => EncounterDiagnosisAdapter.tearDown()],
+    // Adapters registered by mounted forms (src/lifecycle.ts keeps a module-level
+    // Set), plus pageObserver.clear().
+    ['lifecycle.teardown', () => lifecycleTeardown()],
+    // pageObserver.clear() sets evaluatedPagesVisibility to false, but its initial
+    // BehaviorSubject value is null ("not yet evaluated") — restore that.
+    ['pageObserver.evaluatedPagesVisibility', () => pageObserver.setEvaluatedPagesVisibility(null)],
```
[the remark this answers is not in the document]
Comment (ibacher) [reply src/test-support/reset.ts:52]: Fixed at the declaration

File: src/test-support/builders.ts [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    questions = [],
+    pages,
+  } = options;
+  const defaultPageLabel = 'Page 1';
+  const defaultPageId = derivePageId(defaultPageLabel, 0);
+  if (!pages) {
+    questions.forEach((question) => {
+      question.meta = { ...(question.meta ?? {}), pageId: defaultPageId };
+    });
+  }
```
Comment (denniskigen) [inline src/test-support/builders.ts:107]:
```suggestion
  if (!pages) {
    const stamp = (question: FormField) => {
      question.meta = { ...(question.meta ?? {}), pageId: defaultPageId };
      question.questions?.forEach(stamp);
    };
    questions.forEach(stamp);
  }
```
`DefaultFormSchemaTransformer.handleQuestion` recurses into group children before stamping `meta.pageId`, so the real pipeline stamps them too - I checked with a scratch run: the transformer gives a group child `page-Page1-0` while these builders leave it undefined. Without the recursive stamp, `findFieldSection` on a group child works in production but returns undefined against these fixtures.

**Review by denniskigen — COMMENTED** [review]
This is great, @ibacher - one thing on the builders.

**Review by denniskigen — APPROVED** [review]

File: src/test-support/builders.ts (hunk 4 above)
Comment (ibacher) [reply src/test-support/builders.ts:107]: Nice catch!
